### PR TITLE
Add stream reader ReadAllAsync method

### PR DESCRIPTION
**Pull request by JamesNK**
Addresses https://github.com/grpc/grpc-dotnet/issues/41

Adds a ReadAllAsync method that returns `IAsyncEnumerable<T>` and is usable with C# 8's async foreach:

```cs
public override async Task<CounterReply> AccumulateCount(IAsyncStreamReader<CounterRequest> requestStream, ServerCallContext context)
{
    await foreach (var request in requestStream.ReadAllAsync())
    {
        _logger.LogInformation($"Incrementing count by {request.Count}");

        _counter.Increment(request.Count);
    }

    return new CounterReply { Count = _counter.Count };
}
```

This PR builds on top of https://github.com/grpc/grpc/pull/19059 and includes its changes (without its changes the C# compiler would complain that `IAsyncEnumerable<T>` is in multiple assemblies)

New code is here: https://github.com/grpc/grpc/commit/7f7b2ad7a53fd3aed857aa64948992e7a50569e0

TODO: Add unit tests (this will involve adding a 3.0 test target to access the netstandard2.1 target). I want to confirm that this addition is *OK* before adding tests.

// @jtattermusch @JunTaoLuo @davidfowl

### CODE REVIEW

File: src/csharp/Grpc.Core.Api/AsyncStreamReaderExtensions.cs.cs [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#endregion
+
+using System.Collections.Generic;
+using System.Threading;
+
+namespace Grpc.Core
```
Comment (JamesNK) [inline src/csharp/Grpc.Core.Api/AsyncStreamReaderExtensions.cs.cs]: I like that the extension method is immediately available to someone writing a gRPC service via the Grpc.Core namespace. If the namespace was Grpc.Core.Util then they would have to know to add it.

**Comment by JunTaoLuo** [comment]
Why are the tests failing?

**Comment by JamesNK** [comment]
> The current .NET SDK does not support targeting .NET Standard 2.1.  Either target .NET Standard 2.0 or lower, or use a version of the .NET SDK that supports .NET Standard 2.1.

A 3.0 preview version is needed to support netstandard21.

File: src/csharp/Grpc.Core.Api/AsyncStreamReaderExtensions.cs.cs [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+#if GRPC_CSHARP_SUPPORT_ASYNC_ENUMERABLE
+    /// <summary>
+    /// Extension methods for <see cref="IAsyncStreamReader{T}"/>.
+    /// </summary>
+    public static class IAsyncStreamReaderExtensions
+    {
+        /// <summary>
+        /// Read all results from the specified stream reader.
+        /// </summary>
+        public async static IAsyncEnumerable<T> ReadAllAsync<T>(this IAsyncStreamReader<T> streamReader, CancellationToken cancellationToken = default)
```
Comment (davidfowl) [inline src/csharp/Grpc.Core.Api/AsyncStreamReaderExtensions.cs.cs]: The cancellation token should have an [EnumeratorCancellation] attribute. This allows the token to flow through from GetAsyncEnumerator(token) to this method.

**Comment by JamesNK** [comment]
I believe https://github.com/grpc/grpc/blob/818908ae9c9c1b124a6d14bca13513b2e2c1b542/tools/run_tests/helper_scripts/pre_build_csharp.sh needs to be modified to fetch the latest version of .NET Core, similar to what is done in grpc/grpc-dotnet.

I'm not familiar with how this repo is built, or with using bash. @jtattermusch when you have time could you add support for fetching and installing the .NET version referenced in global.json to the build. I will probably make a mess if I tried.

**Comment by JamesNK** [comment]
Implemented for now at https://github.com/grpc/grpc-dotnet/pull/438

**Pull request closed without merging**